### PR TITLE
Batch delete images.

### PR DESCRIPTION
**Pull request by akash1810**
Haven't spent too long on styling as this'll move to the toolbar soon (hopefully!).

screen shot 2015-08-17 at 17 41 51

### CODE REVIEW

**Comment by alastairjardine** [comment]
It does need to move soon…v v soon!

File: kahuna/public/js/components/gr-panel/gr-panel.css [hunk 1]
```diff
@@ -61,26 +61,17 @@
...
 
 .batch-archive {
-    float: left;
     width: 100%;
-    box-sizing: border-box;
-    /* ^ because it has elements floating in it */
-}
-
-.batch-archive .metadata-line {
-    float: left;
```
Comment (akash1810) [inline kahuna/public/js/components/gr-panel/gr-panel.css]: :heart: [flexbox](https://github.com/guardian/grid/pull/1101/files#diff-616e7ea78eaec4cf50c56788e7f9711dR272)

**Comment by akash1810** [comment]
I've released this to TEST now.

File: kahuna/public/js/search/results.js [hunk 2]
```diff
@@ -370,6 +370,23 @@ results.controller('SearchResultsCtrl', [
...
+
+            if (imageIndex > -1) {
+                ctrl.images.splice(imageIndex, 1);
+            }
+
+            if (imageAllIndex > -1) {
+                ctrl.imagesAll.splice(imageIndex, 1);
+            }
+
+            ctrl.totalResults += -1;
```
Comment (theefer) [inline kahuna/public/js/search/results.js]: did you mean `ctrl.totalResults -= 1` or even `ctrl.totalResults--`? :-)

File: kahuna/public/js/search/results.js [hunk 3]
```diff
@@ -370,6 +370,23 @@ results.controller('SearchResultsCtrl', [
             }
         });
 
+        const freeImageDeleteListener = $rootScope.$on('image-deleted', (e, image) => {
```
Comment (theefer) [inline kahuna/public/js/search/results.js]: Bit crappy to have to do this kind of stuff here, but agree it's necessary until we tidy up all this stateful business.

**Comment by theefer** [comment]
Looks good, but I just tried it in TEST and it looks a bit buggy when deleting two consecutive images. The second doesn't disappear.

I've got a suspicion that images don't like their index changing in the lazy table (an assumption was made that the index never changes iirc). One way to avoid this might be not to remove them from the `images` and `imagesAll` arrays, but instead replace them by `undefined`. Although actually now I think about it it would mean they are not loaded and would be reloaded, causing other duplicates, etc.

Let's catch up tomorrow morning, it's probably a reasonably easy fix.

**Comment by akash1810** [comment]
this is on TEST again, it looks ok, if a little slow because of the setTimeout.

File: kahuna/public/js/components/gr-delete-image/gr-delete-image.js [hunk 4]
```diff
@@ -3,23 +3,30 @@ import '../gr-confirm-delete/gr-confirm-delete';
...
+        };
+
+        ctrl.delete = function () {
+            // HACK to wait for thrall to process the message so that when we
+            // poll the api, it will be up to date.
+            return $q.all(ctrl.images.map(image => ctrl.deleteImage(image)))
+                .then(() => {
+                    $timeout(() => {
+                        $rootScope.$emit('images-deleted', ctrl.images);
+                    }, 1000);
```
Comment (akash1810) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: I wonder if we should make the timeout relative to the number of images? `Math.min(ctrl.images.length * 500, 2000)`?

Comment (theefer) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: Why? They should be processed within milliseconds if the queue is consumed normally.

Comment (akash1810) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: There was an image that took a bit longer to delete on TEST - resulting in the crazy disappearing images. A page refresh sorted, but obv not ideal.

**Comment by akash1810** [comment]
Caught up with master now! This means the delete button is in the toolbar, rather than the panel!

**Comment by alastairjardine** [comment]
Sweet

On Friday, 21 August 2015, Akash A <notifications@github.com> wrote:

> Caught up with master now! This means the delete button is in the toolbar,
> rather than the panel!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/guardian/grid/pull/1101#issuecomment-133504804>.
>


-- 
--
Alastair Jardine, Senior UX Architect

-- 

------------------------------
Visit theguardian.com. On your mobile and tablet, download the Guardian 
iPhone and Android apps theguardian.com/guardianapp and our tablet editions 
theguardian.com/editions.  Save up to 57% by subscribing to the Guardian 
and Observer - choose the papers you want and get full digital access. 
 Visit subscribe.theguardian.com

This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396

File: kahuna/public/js/components/gr-delete-image/gr-delete-image.js [hunk 5]
```diff
@@ -3,23 +3,30 @@ import '../gr-confirm-delete/gr-confirm-delete';
...
+            return mediaApi.delete(image)
+                .catch((err) => {
+                    $rootScope.$emit('image-delete-failure', err, image);
+                });
+        };
+
+        ctrl.delete = function () {
+            // HACK to wait for thrall to process the message so that when we
+            // poll the api, it will be up to date.
+            return $q.all(Array.from(ctrl.images.values()).map(image => ctrl.deleteImage(image)))
```
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: Why not use the polling mechanic we have else where for checking when thralls done?
Or look into something like [`SSE`s](http://www.html5rocks.com/en/tutorials/eventsource/basics/)?

Comment (theefer) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: I don't like all this polling we're doing, but given how much pain we're having working around issues with this, that might be the best/simplest/mose reliable idea.. @akash1810 what do you think?

Comment (akash1810) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: I'm tempted to try SSEs though am conscious this PR has been open (and wip-ed) for too long, so will time box it.

Comment (theefer) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: Can we try and deliver this feature first? I think @jamesgorrie's suggestion could help fix the current issue blocking release and seems sensible to me.

Not sure how we'd implement SSE in this case without architectural changes, but could be interesting to chat about it when @jamesgorrie is back in so we have an idea if and when it fits on the roadmap.

Comment (akash1810) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: updated w/polling

Comment (theefer) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: Nice one!

**Comment by theefer** [comment]
:+1: great

File: kahuna/public/js/components/gr-delete-image/gr-delete-image.js (hunk 5 above)
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: Yea, also didn't want to splinter off onto another way of doing things, as this is starting to complicate the code (having multiple ways of doing things). If we'd like to we could abstract it to a checkActionComplete service or something and then muck around with the complexities and different implementations in there to keep it in one place... Cheers @akash1810.

**Comment by jamesgorrie** [comment]
bonza :+1:

**Pull request merged**